Code/GraphMol/ChemReactions/ReactionRunner.cpp

```diff
@@ -854,4 +882,192 @@ std::vector<MOL_SPTR_VECT> run_Reactants(const ChemicalReaction &rxn,
   return productMols;
 }  // end of ChemicalReaction::runReactants()
 
+// Generate a product based on a SINGLE reactant
+std::vector<MOL_SPTR_VECT> run_Reactant(const ChemicalReaction &rxn,
+                                        const ROMOL_SPTR &reactant,
+                                        unsigned int reactantIdx) {
+  if (!rxn.isInitialized()) {
+    throw ChemicalReactionException(
+        "initMatchers() must be called before runReactants()");
+  }
+
+  CHECK_INVARIANT(reactant, "bad molecule in reactants");
+  reactant->clearAllAtomBookmarks();  // we use this as scratch space
+  std::vector<MOL_SPTR_VECT> productMols;
+
+  // if we have no products, return now:
+  if (!rxn.getNumProductTemplates()) {
+    return productMols;
+  }
+
+  CHECK_INVARIANT(static_cast<size_t>(reactantIdx) < rxn.getReactants().size(),
+                  "reactantIdx out of bounds");
+  // find the matches for each reactant:
+  VectVectMatchVectType matchesByReactant;
+
+  // assemble the reactants (use an empty mol for missing reactants)
+  MOL_SPTR_VECT reactants(rxn.getNumReactantTemplates());
+  for (size_t i = 0; i < rxn.getNumReactantTemplates(); ++i) {
+    if (i == reactantIdx)
+      reactants[i] = reactant;
+    else
+      reactants[i] = ROMOL_SPTR(new ROMol);
+  }
+
+  if (!ReactionRunnerUtils::getReactantMatches(
+          reactants, rxn, matchesByReactant, reactantIdx)) {
+    return productMols;
+  }
+
+  VectMatchVectType &matches = matchesByReactant[reactantIdx];
+  // each match on a reactant is a seperate product
+  VectVectMatchVectType matchesAtReactants(matches.size());
+  for (size_t i = 0; i < matches.size(); ++i) {
+    matchesAtReactants[i].resize(rxn.getReactants().size());
+    matchesAtReactants[i][reactantIdx] = matches[i];
+  }
+  productMols.resize(matches.size());
+
+  for (unsigned int productId = 0; productId != productMols.size();
+       ++productId) {
+    MOL_SPTR_VECT lProds = ReactionRunnerUtils::generateOneProductSet(
+        rxn, reactants, matchesAtReactants[productId]);
+    productMols[productId] = lProds;
+  }
+
+  return productMols;
+}  // end of ChemicalReaction::runReactants()
+
+namespace {
+int getAtomMapNo(ROMol::ATOM_BOOKMARK_MAP *map, Atom *atom) {
+  if (map) {
+    for (ROMol::ATOM_BOOKMARK_MAP::const_iterator it = map->begin();
+         it != map->end(); ++it) {
+      for (ROMol::ATOM_PTR_LIST::const_iterator ait = it->second.begin();
+           ait != it->second.end(); ++ait) {
+        if (*ait == atom) return it->first;
+      }
+    }
+  }
+  return -1;
+}
+}
+
+namespace {
+struct RGroup {
+  Atom *rAtom;
+  Bond::BondType bond_type;
+  int mapno;
+  RGroup(Atom *atom, Bond::BondType type, int curmapno = -1)
+      : rAtom(atom), bond_type(type), mapno(curmapno) {}
+  RGroup(const RGroup &rhs)
+      : rAtom(rhs.rAtom), bond_type(rhs.bond_type), mapno(rhs.mapno) {}
+};
+}
+ROMol *reduceProductToSideChains(const ROMOL_SPTR &product, bool addDummyAtoms) {
```

Comment:
You should probably mention in the docstring properties should be set on the product. Otherwise this functions looks fantastically magical from a first glance at the API :)